### ISSUE TITLE
[QUERY] Azure.Storage.Files.DataLake on behalf of user authentication

### ISSUE THREAD

**Issue by fariskarcic** [Azure/azure-sdk-for-net#9490]
**Query/Question**
Issue I'm having with this sdk is that I cant figure out how to authN user against my datalake. There are so many options on TokenCredentials, but none to generate new token from existing one. Went through documentation and could not figure out. I'm trying to achieve this on WebAPI where I already have access token for my own system from AD, authenticated for this app in my tenant.

***Why is this not a Bug or a feature Request?***
I'm unsure if there is way to do this.

**Setup (please complete the following information if applicable):**
 - Running .NET Core 3.1
 - Azure.Storage.Files.DataLake 12.0.0 preview 8

**Comment by wantedfast** [comment]
Hi ,@fariskarcic

**Comment by wantedfast** [comment]
We provide several options of authenticating user. You can find more information in the below document. 
https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/identity/Azure.Identity
You are recommended to use the DefaulAzureCredential since you already have an AD. The detailed steps are as following:
	1. You need to first configure the environment setting the variables AZURE_CLIENT_ID, AZURE_TENANT_ID and AZURE_CLIENT_SECRET to the appropriate values for your service principal.
	2. Using DefaultAzureCredential to create a DataLakeServiceClient instance.
DataLakeServiceClient serviceClient = new DataLakeServiceClient(new Uri("<YourServiceUri>"), new DefaultAzureCredential());

**Comment by fariskarcic** [comment]
Yes, I've done it this way also, but my specific case requires ti know about current user.

**Comment by seanmcc-msft** [comment]
@fariskarcic, can you tell us more about your use case?

**Comment by fariskarcic** [comment]
@seanmcc-msft 
Goes like this, users have their own workspaces on data lake, creation of which is handled by administrator (not handled by my app). They only have permissions over their workspaces, and can just see and work with them. My app has multiple authentications/authorizations with AAD for logging in, graph, third party api, and for data lake. Everything is handled by "on behalf of user" flow.

So whole app is centered around user interacting with various systems, which is being logged and audited.

**Comment by nadirbad** [comment]
Hi @seanmcc-msft, @wantedfast . Me and @fariskarcic are members of the team working on this project. Here is a more detailed description about our use case.

We have a web API with Azure Ad authentication using MSAL. Our API uses on-behalf-of authentication flow to access other backend APIs, Microsoft Graph. Basically we need to authorize access to Storage account (gen2) on the user's behalf, meaning resources are accessed using the permissions of the user who is logged in. 
Using old Microsoft.Azure.Storage package this was possible, here is the documentation [link](https://docs.microsoft.com/en-us/azure/storage/common/storage-auth-aad-app?toc=%2fazure%2fstorage%2fblobs%2ftoc.json#create-a-web-application-that-authorizes-access-to-blob-storage-with-azure-ad).

How can we authenticate to storage account, when we only have token of current user?
Basically we need something like in this [example](https://github.com/Azure-Samples/storage-dotnet-azure-ad-msal/blob/master/Controllers/HomeController.cs) or a way to use user's token with the new Azure.Core.TokenCredential.

Any help is appreciated. Thank you.

**Comment by seanmcc-msft** [comment]
Hi @nadirbad, We have constructors taking ```Azure.Core.TokenCredential``` on all our Data Lake clients (```DataLakeServiceClient```, ```DataLakeFileSystemClient```, ```DataLakeFileClient```, etc.).  To construct a ```TokenCredential```, I'd recommend taking a look at https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/identity/Azure.Identity.

**Comment by EdwinStraver** [comment]
@nadirbad @fariskarcic I have the exact same issue; were you able to resolve this?

**Comment by jongio** [comment]
@nadirbad - If you want to use a token that you already have from user auth, you have to implement your own credential type that derives from TokenCredential.  If you do so, you should also implement the caching and refresh of the token.

@schaabs - Do you have a .NET version of this that we can share?

**Comment by schaabs** [comment]
@nadirbad Sorry you're having this trouble. We're currently working on a set of samples which demonstrate different implementation of `TokenCredential` for cases where the Azure.Identity library is not sufficient for an application. While we haven't published these yet I've shared the two which are relevant here in a gist.

The first example, [Authenticating with a prefetched access token](https://gist.github.com/schaabs/edbce217621e5ce5aca1a8fa6f711952#authenticating-with-a-prefetched-access-token), demonstrates creating a `TokenCredenial` implementation which is constructed with a static access token. This is similar to the approach taken in the sample you shared.

Also, since you're using the on behalf of authentication flow, I included a second example, [Authenticating with the On Behalf Of Flow](https://gist.github.com/schaabs/edbce217621e5ce5aca1a8fa6f711952#authenticating-with-the-on-behalf-of-flow). This demonstrates defining a custom `TokenCredential` which uses a MSAL `IConfidentialClientApplication` to issue obo tokens.

Hopefully these samples help to unblock you. If you have any questions or trouble getting them to work please let me know.

**Comment by nadirbad** [comment]
@schaabs Thank you for examples. We managed to overcome this "blocker" by moving this piece of logic to our front-end application (React) and using Data Lake Storage Gen2 REST API instead.

**Comment by navba-MSFT** [comment]
@fariskarcic Apologies for the late reply. I just wanted to check if you had a chance to look at the above comment provided by @schaabs. Please let us know if you have any pending questions. We would be happy to address.
